File: bench/dcn/ranking.py
"""The learned order, ported from site/js/ranking.js.

The weights come from learn.py and live in site/taxonomy/ranking.json. The
benchmark has to rank models exactly as the page does, or it would be scoring
advice nobody is given, so this mirrors the JavaScript and the agreement test
compares the two.

Three kinds of order can be in use ("chosen" in ranking.json):

  prior       one number per model: its average percentile rank
  prior_fit   the prior plus fitted interactions with the dataset's features
  prior_knn   the prior blended with what the model was worth on the
              benchmark datasets nearest to this one

The blend for prior_knn is (S * prior + sum w * rank) / (S + sum w) over the k
nearest benchmark datasets, where rank is the model's percentile rank on that
dataset and w = exp(-(distance / h)^2). Close neighbours pull the score toward
what happened on them; far ones barely move it off the prior. S, k and h are
fixed in learn.py before anything is evaluated, not tuned on the results.
"""
from __future__ import annotations

import json
import math
from pathlib import Path

ROOT = Path(__file__).resolve().parents[2]
DEFAULT_PRIOR = 0.5
FLAGS = ["A31", "A32", "A38", "A41", "A42", "A43", "A51", "A52", "A53", "A54", "A61", "A62", "A64"]


def load_ranking(root: Path = ROOT) -> dict | None:
    path = root / "site" / "taxonomy" / "ranking.json"
    return json.loads(path.read_text()) if path.exists() else None


def ranking_features(sig: dict) -> dict[str, float]:
    codes = set(sig["codes"])
    rows = max(sig.get("rows") or 0, 1)
    features = max(sig.get("features") or 0, 1)
    out = {
        "log_rows": math.log10(rows),
        "log_features": math.log10(features),
        "features_per_row": math.log10(features / rows),
    }
    for code in FLAGS:
        out[f"is_{code}"] = 1.0 if code in codes else 0.0
    return out


def has_learned_ranking(ranking: dict | None, task: str) -> bool:
    return bool(ranking and ranking.get("tasks", {}).get(task))


def meta_vector(meta: dict, names: list[str]) -> list[float]:
    """A dataset's meta-features in a fixed order, at the four decimals both sides store."""
    return [round(float(meta.get(name) or 0.0), 4) for name in names]


def meta_distance(x: list[float], y: list[float], names: list[str], scale: dict) -> float:
    """Mirrors metaDistance() in site/js/ranking.js and distance() in site/js/nearest.js."""
    total = 0.0
    for i, name in enumerate(names):
        spread = scale.get(name) or 1.0
        diff = (x[i] - y[i]) / spread
        total += diff * diff
    return math.sqrt(total / len(names))


def neighbours_of(block: dict, meta: dict) -> list[tuple[dict, float]]:
    """The k nearest benchmark datasets and their kernel weights, nearest first."""
    names, scale = block["features"], block["scale"]
    x = meta_vector(meta, names)
    scored = [(meta_distance(x, d["meta"], names, scale), d["dataset"], d) for d in block["datasets"]]
    scored.sort(key=lambda item: (item[0], item[1]))
    h = block["bandwidth"]
    return [(d, math.exp(-((dist / h) ** 2))) for dist, _, d in scored[:block["k"]]]


def ranking_neighbours(ranking: dict | None, task: str, meta: dict | None):
    """Neighbours for the order in use, or None when it does not use them."""
    task_weights = (ranking or {}).get("tasks", {}).get(task)
    block = (task_weights or {}).get("neighbours")
    if not block or meta is None or ranking.get("chosen") != "prior_knn":
        return None
    return neighbours_of(block, meta)


def blend(prior: float, code: str, neighbours: list[tuple[dict, float]], strength: float) -> float:
    total = weight = 0.0
    for d, w in neighbours:
        rank = d["ranks"].get(code)
        if rank is None:
            continue
        total += w * rank
        weight += w
    return (strength * prior + total) / (strength + weight)


def learned_score(ranking: dict | None, model: dict, task: str, features: dict,
                  neighbours: list | None = None) -> float | None:
    task_weights = (ranking or {}).get("tasks", {}).get(task)
    if not task_weights:
        return None
    prior = task_weights["prior"].get(model["c"])
    if prior is None:
        return None
    if neighbours is not None:
        return blend(prior, model["c"], neighbours, task_weights["neighbours"]["strength"])
    score = prior
    family = (ranking.get("families") or {}).get(model["c"])
    for name, value in features.items():
        score += task_weights.get("weights", {}).get(f"{family}|{name}", 0.0) * value
    return score

File: site/js/ranking.js
// The order models are shown in.
//
// Counting matched coordinates ranked models by how describable they were:
// Linear Regression came first on 17 of 20 regression datasets and cost up to
// 0.7 R². This orders them by what they were actually worth on the benchmark,
// using weights fitted in bench/dcn/learn.py and judged leave-one-dataset-out.
//
// The base is a prior per model: its average percentile rank among the models
// that ran on the same dataset, shrunk toward the middle so a model seen on
// few datasets is not trusted too far. Interactions between the dataset's
// features and the model's family (`weights`) and a blend with the nearest
// benchmark datasets (`neighbours`) are the two richer orders learn.py can
// choose; each is read here, so switching needs no code change.
//
// A task the benchmark never covered (grouping, anomalies, forecasting) has no
// weights, and the ranking falls back to counting coordinates. The page says
// which of the two it used.
//
// When ranking.json's "chosen" is "prior_knn", the prior is blended with what
// each model was worth on the benchmark datasets nearest to the upload:
// (S * prior + sum w * rank) / (S + sum w) over the k nearest, with
// w = exp(-(distance / h)^2). learn.py fixes S, k and h before evaluating, and
// only ships this order if it beats the plain prior by more than luck.
// Mirrors bench/dcn/ranking.py; the agreement test holds the two together.

export const DEFAULT_PRIOR = 0.5;

// Everything the weights can use, computed from what the page already measured.
export function rankingFeatures(sig) {
  const codes = new Set(sig.codes);
  const rows = Math.max(sig.rows ?? 0, 1);
  const features = Math.max(sig.features ?? 0, 1);
  const out = {
    log_rows: Math.log10(rows),
    log_features: Math.log10(features),
    features_per_row: Math.log10(features / rows),
  };
  for (const code of ['A31', 'A32', 'A38', 'A41', 'A42', 'A43', 'A51', 'A52', 'A53', 'A54', 'A61', 'A62', 'A64']) {
    out[`is_${code}`] = codes.has(code) ? 1 : 0;
  }
  return out;
}

// A dataset's meta-features in a fixed order, at the four decimals the
// benchmark stores, so the page and the benchmark measure the same distance.
export function metaVector(meta, names) {
  return names.map(name => Math.round((meta?.[name] ?? 0) * 1e4) / 1e4);
}

export function metaDistance(x, y, names, scale) {
  let total = 0;
  for (let i = 0; i < names.length; i++) {
    const spread = scale?.[names[i]] || 1;
    const diff = (x[i] - y[i]) / spread;
    total += diff * diff;
  }
  return Math.sqrt(total / names.length);
}

// The k nearest benchmark datasets and their weights, nearest first.
export function neighboursOf(block, meta) {
  const x = metaVector(meta, block.features);
  return block.datasets
    .map(d => ({ d, dist: metaDistance(x, d.meta, block.features, block.scale) }))
    .sort((a, b) => a.dist - b.dist || (a.d.dataset < b.d.dataset ? -1 : a.d.dataset > b.d.dataset ? 1 : 0))
    .slice(0, block.k)
    .map(({ d, dist }) => ({ d, w: Math.exp(-((dist / block.bandwidth) ** 2)) }));
}

// Neighbours for the order in use, or null when it does not use them.
export function rankingNeighbours(T, task, meta) {
  const block = T.RANKING?.tasks?.[task]?.neighbours;
  if (!block || !meta || T.RANKING.chosen !== 'prior_knn') return null;
  return neighboursOf(block, meta);
}

export function blend(prior, code, neighbours, strength) {
  let total = 0, weight = 0;
  for (const { d, w } of neighbours) {
    const rank = d.ranks[code];
    if (rank == null) continue;
    total += w * rank;
    weight += w;
  }
  return (strength * prior + total) / (strength + weight);
}

export function hasLearnedRanking(T, task) {
  return Boolean(T.RANKING?.tasks?.[task]);
}

// A model's learned score, or null when this task was never benchmarked.
export function learnedScore(T, model, task, features, neighbours = null) {
  const ranking = T.RANKING?.tasks?.[task];
  if (!ranking) return null;
  const prior = ranking.prior[model.c];
  if (prior == null) return null;   // never benchmarked: no claim to make
  if (neighbours) return blend(prior, model.c, neighbours, ranking.neighbours.strength);
  let score = prior;
  const family = T.RANKING.families?.[model.c];
  for (const [name, value] of Object.entries(features)) {
    score += (ranking.weights?.[`${family}|${name}`] ?? 0) * value;
  }
  return score;
}

// How many datasets the weights were fitted on, for the page to state.
export function rankingProvenance(T, task) {
  const ranking = T.RANKING?.tasks?.[task];
  if (!ranking) return null;
  return {
    datasets: ranking.datasets,
    trainedOn: T.RANKING.trained_on,
    chosen: T.RANKING.chosen,
  };
}
